consensus/dolev_strong.go

```diff
@@ -0,0 +1,310 @@
+		return false
+	}
+	return out
+}
+
+func (impl *DolevStrongConsensus) getRound() int32 {
+	return int32(impl.timer.Since(impl.startTime).Nanoseconds() / impl.dsConfig.PhaseTime.Nanoseconds())
+}
+
+//GetOutput Returns the message agreed upon in this instance of dolev string, returns nil if the message agreement failed
+func (dsci *DolevStrongInstance) GetOutput() []byte {
+	if dsci.abortProcessingInstance {
+		return nil
+	}
+	return dsci.value
+}
+
+// ReceiveMessage is the main receiver handler for a message received from a dolev strong instance running in the network
+func (dsci *DolevStrongInstance) ReceiveMessage(message *pb.ConsensusMessage) error{
+	//calculates round according to local clock
+	round := dsci.ds.getRound()
+	log.Info("received message in round %v num of signatures: %v", round, len(message.Validators))
+
+	if dsci.abortProcessingInstance {
+		return fmt.Errorf("message received on aborted isntance, round: %v", round)
+	}
+
+	if round > dsci.ds.dsConfig.NumOfAdverseries+1 {
+		return fmt.Errorf("round out of order: %v", round)
+	}
+
+	publicKeys, err := dsci.findAndValidateSignatures(message)
+	if err != nil {
+		return err
+	}
+	initiatorPubKey, err := crypto.NewPublicKey(message.Msg.AuthPubKey)
+	if _, ok := publicKeys[initiatorPubKey.String()]; !ok || err != nil {
+		return fmt.Errorf("initiator did not sign the message - aborting")
+
+	}
+	//check if there are enough signature to match round number
+	if len(publicKeys) < int(round) {
+		return fmt.Errorf("invalid number of signatures - not matching round number %v num of signatures %v", round, len(publicKeys))
+	}
+	//todo: what if the signatures are of nodes that are not in the layer?
+	if dsci.value != nil {
+		if res := bytes.Compare(dsci.value, message.Msg.Data); res != 0 {
+			dsci.abortProcessingInstance = true //with abort
+			return fmt.Errorf("received two different messages from same sender, aborting this instance")
+		}
+		// no need to do anything
+		return nil
+	}
+	//set the message as the correct one
+	dsci.value = message.Msg.Data
+	log.Info("Message received on node : %v", dsci.ds.publicKey.String())
+	// add our signature on the message
+	err = dsci.ds.signMessageAndAppend(message)
+	if err != nil {
+		return fmt.Errorf("cannot sign message %v error %v", message, err)
+	}
+	//add this node to the nodes already signed
+	publicKeys[dsci.ds.publicKey.String()] = struct{}{}
+	dsci.ds.sendToRemainingNodes(message, publicKeys) // to be sent in round +1
+	return nil
+}
+
+func (dsci *DolevStrongInstance) findAndValidateSignatures(message *pb.ConsensusMessage) (map[string]struct{}, error){
+	publicKeys := make(map[string]struct{})
+	// validate signatures
+	for _, val := range message.Validators {
+		key, ok := crypto.NewPublicKey(val.AuthPubKey)
+		if ok != nil {
+			return nil, errors.New("cannot read validator public key")
+		}
+
+		validSig := dsci.authenticateSignature(key, val.AuthorSign, message.Msg)
```

Comment:
authenticateSignature will marshal message.Msg for every validator. Consider marshaling in advance and sending the raw bytes instead of the message